### DISCUSSION TITLE
When two bodies are connected with a weld joint, translating one of the bodies does not work correctly.

### DISCUSSION THREAD

**Discussion by riftek** [dyn4j/dyn4j#277] [Q&A]
I have a method in the body that when its minY reaches the end of the worlds bounds its shifted by the length of the bounds.
When 2 bodies are connected via a weld joint, the sub bodies mainBody is set to  false so that it does not shift the sub body.
But when I Shift the main body, it seems to not shift the full bounds size 800 px. It shifts fine without the weld joint.
it seems to do it by exactly the height of the sub body

**Comment by wnbittle** [comment]
Can you provide the coordinates of the main body before the translation, right after the translation, and on the next few timesteps?  I think this would help track down the issue.

**Comment by riftek** [comment]
Heres without the weld joint.
start: (200.0, 233.3333333333333)
translate 800: (200.0, 1033.3333333333333)
apply force down

process detcting minX>worldSize
before: (200.0, 1033.333449074074)
After: (200.0, 233.33344907407403)

here is with 2 bodies welded.
start: (200.0, 233.3333333333333)
translate 800: (200.0, 1033.3333333333333)
it is not detected that minX>worldSize
in update method
center: (200.0, 833.3334201388889)

during the update something changes its position

@Override
    public void update() {
        updatePositioningInBounds();

    }
    
    I dont have anything that could be changing the position before the update

**Comment by wnbittle** [comment]
Hmmm, can you try turning off CCD (continuous collision detection) in `Settings`?

**Reply by riftek** [reply]
Unfortunately the same thing

**Comment by wnbittle** [comment]
All good, just trying to eliminate sub-systems.  Could you give me a screenshot of the simulation before and after the translation?  Another question, could you describe _when_ you are translating the body across the screen?  For example, are you doing it like the following, or are you doing it a different way:

```java
world.update(...);
// move the body
if (body.minY < world.bounds.y) {
  body.translate(0, ...);
}
```

My guess right now is that the `WeldJoint`'s the position constraint solver is causing this issue.

**Reply by riftek** [reply]
yeah Im using body.translate(0,y)
this is before
![image](https://user-images.githubusercontent.com/30284512/216622649-7399fe31-3a5d-4e7c-8e0f-5163559ff947.png)

its supposed to be off the screen, but it gets shifted.
![image](https://user-images.githubusercontent.com/30284512/216622188-29fcdfb6-411d-45cb-8d56-451dfeb71f84.png)

**Reply by wnbittle** [reply]
Thank you for the screenshots.  Regarding how you translate, my question was **_when_** you call `Body.translate`.  Do you call it after the `world.update` method completes, do you call it from a `Listener`, etc.?

**Reply by riftek** [reply]
I call body.translate in the constructor when its initialised and added to the world.. before any updates.

**Reply by riftek** [reply]
Then in body.update has minY>worldSize..but that statement is false after translating 800

**Comment by wnbittle** [comment]
Another question - the two bodies in question, they are both `MassType.NORMAL`?  Can you share the body specs (coordinates, mass, inertia, etc.) as well?

One thing you can try is in this scenario destroy and rebuild the joint:
```java
if (out of bounds) {
  b.translate(...);
  w.removeJoint(...);
  WeldJoint wj = new WeldJoint(...);
  w.addJoint(...);
}
```

I'm pretty sure it's the position solver for the `WeldJoint` is the "problem", but I'm afraid that it's working as expected.

**Reply by riftek** [reply]
Thanks, are you able to replicate the issue?

**Reply by wnbittle** [reply]
> Thanks, are you able to replicate the issue?

Yes

My simple reproduction case is:
```java
World w = ...;
w.setGravity(World.ZERO_GRAVITY);

Body b1 = ...;
b1.setMass(MassType.NORMAL);
Body b2 = ...;
b2.setMass(MassType.NORMAL);

WeldJoint wj = ...;

w.addBody(b1);
w.addBody(b2);
w.addJoint(wj);

b1.translate(100, 0);
w.step(1);

// assuming both bodies are the same mass
// b1's location is now (50, 0)
// b2's location is now (50, 0)
```

What's happening is that the `WeldJoint` is seeing the translation of _one_ of the bodies as constraint _drift_ and the position solver comes in and "solves" the drift.  In my case, since both bodies have the same mass, they are moved an equal distance.

As the velocity solver continues to keep the bodies constrained by applying impulses (which apply to the velocity), over time the bodies will drift positionally.  The position solver comes in to resolve the drift which results in stable rigid constraints.  Unfortunately, it's working against you here.  That said, direct position manipulation isn't really recommended and there could be other issues that arise.

**Reply by riftek** [reply]
Thanks! I think I get it now.

**Comment by riftek** [comment]
Im not sure why I didnt try this earlier, if you shift all the shapes it works as expected.
